### ISSUE TITLE
It is unclear how you should return objects in a rust COM InProc Server

### ISSUE THREAD

**Issue by rbrownwsws** [microsoft/windows-rs#1819]
I'm trying to make a COM InProc server (for a [shell namespace extension](https://docs.microsoft.com/en-us/windows/win32/shell/nse-works)) using [this C++ sample](https://github.com/microsoft/Windows-classic-samples/tree/main/Samples/Win7Samples/winui/shell/shellextensibility/explorerdataprovider) as a guide.

I've run into a problem implementing [`DllGetClassObject()`](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-dllgetclassobject) and I am not sure if it is because I am using the windows-rs COM support wrong or if there is a problem elsewhere with the logic of my program.

From my logs I can see windows call my `DllGetClassObject()` method repeatedly but it then never calls `IClassFactory::CreateInstance()` on any of the objects I return.
This makes me suspicious that I am doing something wrong with my COM object so windows cannot use the pointer I return.
Can you tell me what I have done wrong or is there an example of using windows-rs in this way I can look at?

I have put the full source in the collapsible sections at the bottom but here are some snippets of what I think are the relevant parts of my code:
```rust
#[windows::core::implement(IClassFactory)]
pub struct RgbzipClassFactory {
// <snip...>

impl IClassFactory_Impl for RgbzipClassFactory {
    fn CreateInstance(
        &self,
        punkouter: &Option<windows::core::IUnknown>,
        riid: *const GUID,
        ppvobject: *mut *mut c_void,
    ) -> ::windows::core::Result<()> {
// <snip...>

pub extern "stdcall" fn DllGetClassObject(
    rclsid: *const GUID,
    riid: *const GUID,
    ppv: *mut *mut c_void,
) -> HRESULT {
  // <snip...>
  
  let factory = RgbzipClassFactory::new(DLL_REF_COUNT.clone());
  let factory: IClassFactory = factory.into();
  
  // Should I be doing this?
  // If I do not `factory` seems to be dropped immediately.
  // If I do `factory` seems to never be dropped but that could be because I somehow I mangled
  // the COM interface and windows also cannot call `IUnknown::Release()`.
  let factory = ManuallyDrop::new(factory);
  
  // [out] LPVOID   *ppv
  *ppv = factory.as_raw();
  
  S_OK
}
```

<details>
<summary>lib.rs</summary>

```rust
mod class_factory;
mod shell_folder;
mod utils;

use core::ffi::c_void;
use std::mem::ManuallyDrop;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use lazy_static::lazy_static;

use log::{debug, warn, LevelFilter};
use log4rs::append::file::FileAppender;
use log4rs::config::{Appender, Root};
use log4rs::Config;

use windows::core::{Interface, GUID, HRESULT};
use windows::Win32::Foundation::{
    BOOL, CLASS_E_CLASSNOTAVAILABLE, E_UNEXPECTED, HINSTANCE, S_FALSE, S_OK,
};
use windows::Win32::System::Com::IClassFactory;
use windows::Win32::System::LibraryLoader::DisableThreadLibraryCalls;
use windows::Win32::System::SystemServices::DLL_PROCESS_ATTACH;

use crate::class_factory::RgbzipClassFactory;
use crate::utils::guid_to_string;

// {C1183A9B-F5C0-4DF3-9C29-709AFDFD7FB2}
pub const CLSID_RGBZIP: GUID = GUID::from_u128(0xC1183A9BF5C04DF39C29709AFDFD7FB2);

// IClassFactory
// {00000001-0000-0000-C000-000000000046}
const IID_I_CLASS_FACTORY: GUID = GUID::from_u128(0x0000000100000000C000000000000046);

lazy_static! {
    /// The number of references to this module (DLL)
    static ref DLL_REF_COUNT: Arc<AtomicUsize> = Arc::new(AtomicUsize::new(0));
}

/// See [`DllMain()`](https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain).
#[no_mangle]
pub extern "stdcall" fn DllMain(hinst: HINSTANCE, reason: u32, _reserved: *mut c_void) -> BOOL {
    if reason == DLL_PROCESS_ATTACH {
        // Set up global logger
        let default_appender = FileAppender::builder()
            .build(r#"C:\Windows\Temp\rgbzip-debug.log"#)
            .unwrap();
        let config = Config::builder()
            .appender(Appender::builder().build("default", Box::new(default_appender)))
            .build(
                Root::builder()
                    .appender("default")
                    .build(LevelFilter::Trace),
            )
            .unwrap();
        log4rs::init_config(config).expect("failed to init logger");

        debug!("DllMain: DLL_PROCESS_ATTACH");

        // Optimisation copied from "explorerdataprovider" sample
        unsafe { DisableThreadLibraryCalls(hinst) };
    }

    BOOL::from(true)
}

/// See [`DllCanUnloadNow()`](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-dllcanunloadnow).
#[no_mangle]
pub extern "stdcall" fn DllCanUnloadNow() -> HRESULT {
    let ref_count = DLL_REF_COUNT.load(Ordering::SeqCst);

    debug!("DllCanUnloadNow: refs: {}", ref_count);

    if ref_count == 0 {
        S_OK
    } else {
        S_FALSE
    }
}

/// See [`DllGetClassObject()`](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-dllgetclassobject).
#[no_mangle]
pub extern "stdcall" fn DllGetClassObject(
    rclsid: *const GUID,
    riid: *const GUID,
    ppv: *mut *mut c_void,
) -> HRESULT {
    debug!("DllGetClassObject: Called");

    let rclsid = &unsafe { *rclsid };
    let riid = &unsafe { *riid };
    let ppv = &mut unsafe { *ppv };

    // ppv must be null if we fail so set it here for safety
    *ppv = std::ptr::null_mut();

    if *rclsid != CLSID_RGBZIP {
        warn!(
            "DllGetClassObject: Unexpected CLSID requested: {}",
            guid_to_string(rclsid),
        );

        return CLASS_E_CLASSNOTAVAILABLE;
    }

    if *riid != IID_I_CLASS_FACTORY {
        warn!(
            "DllGetClassObject: Unexpected IID requested: {}",
            guid_to_string(riid),
        );

        return E_UNEXPECTED;
    }

    let factory = RgbzipClassFactory::new(DLL_REF_COUNT.clone());
    let factory: IClassFactory = factory.into();
    let factory = ManuallyDrop::new(factory);
    *ppv = factory.as_raw();

    debug!("DllGetClassObject: Success: {:p}", *ppv);

    S_OK
}

```

</details>

<details>
<summary>class_factory.rs</summary>

```rust
use core::ffi::c_void;
use std::mem::ManuallyDrop;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use log::{debug, warn};

use windows::core::{Interface, GUID};
use windows::Win32::Foundation::{BOOL, CLASS_E_NOAGGREGATION, E_NOINTERFACE};
use windows::Win32::System::Com::{IClassFactory, IClassFactory_Impl};
use windows::Win32::UI::Shell::IShellFolder;

use crate::shell_folder::RgbzipShellFolder;
use crate::utils::guid_to_string;

// IShellFolder
// {000214E6-0000-0000-C000-000000000046}
const IID_I_SHELL_FOLDER: GUID = GUID::from_u128(0x000214E600000000C000000000000046);

#[windows::core::implement(IClassFactory)]
pub struct RgbzipClassFactory {
    dll_ref_count: Arc<AtomicUsize>,
}

impl RgbzipClassFactory {
    pub fn new(dll_ref_count: Arc<AtomicUsize>) -> Self {
        let old_dll_refs = dll_ref_count.fetch_add(1, Ordering::SeqCst);

        debug!("Created ClassFactory - {} refs", old_dll_refs + 1);

        RgbzipClassFactory { dll_ref_count }
    }
}

impl Drop for RgbzipClassFactory {
    fn drop(&mut self) {
        let old_dll_refs = self.dll_ref_count.fetch_sub(1, Ordering::SeqCst);
        debug!("Dropped ClassFactory - {} refs", old_dll_refs - 1);
    }
}

#[allow(non_snake_case)]
impl IClassFactory_Impl for RgbzipClassFactory {
    fn CreateInstance(
        &self,
        punkouter: &Option<windows::core::IUnknown>,
        riid: *const GUID,
        ppvobject: *mut *mut c_void,
    ) -> ::windows::core::Result<()> {
        let riid = &unsafe { *riid };
        let ppvobject = &mut unsafe { *ppvobject };

        debug!("Tried to create instance of: {}", guid_to_string(riid));

        *ppvobject = std::ptr::null_mut();

        // We do not support aggregation
        if punkouter.is_some() {
            return Err(windows::core::Error::from(CLASS_E_NOAGGREGATION));
        }

        if *riid != IID_I_SHELL_FOLDER {
            warn!(
                "RgbzipClassFactory: Unexpected IID requested: {}",
                guid_to_string(riid),
            );

            return Err(windows::core::Error::from(E_NOINTERFACE));
        }

        let shell_folder = RgbzipShellFolder::new(self.dll_ref_count.clone());
        let shell_folder: IShellFolder = shell_folder.into();
        let shell_folder = ManuallyDrop::new(shell_folder);
        *ppvobject = shell_folder.as_raw();

        Ok(())
    }

    fn LockServer(&self, flock: BOOL) -> ::windows::core::Result<()> {
        debug!("Lock server: {}", flock.as_bool());

        if flock.as_bool() {
            self.dll_ref_count.fetch_add(1, Ordering::SeqCst);
        } else {
            self.dll_ref_count.fetch_sub(1, Ordering::SeqCst);
        }

        Ok(())
    }
}

```

</details>

<details>
<summary>installer.wxs</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi">
    <Product Name="rgbzip"
             Manufacturer="Ross Brown"
             Version="0.1.0"
             Id="*"
             UpgradeCode="40E6CEA0-88F9-405B-A8BF-5905CDCC13EE"
             Language="2057"
             Codepage="1252">
        <Package Id="*"
                 Keywords="Installer"
                 Description="rgbzip Installer"
                 Manufacturer="Ross Brown"
                 Compressed="yes"
                 InstallScope="perMachine"
                 Languages="2057"
                 SummaryCodepage="1252"
                 InstallerVersion="200"/>

        <MajorUpgrade DowngradeErrorMessage="A later version of [ProductName] is already installed."/>

        <MediaTemplate EmbedCab="yes"/>

        <Directory Id="TARGETDIR" Name="SourceDir">
            <Directory Id="ProgramFiles64Folder">
                <Directory Id="INSTALLDIR" Name="rgbzip">
                    <Component Id="ShellExtensionDll" Guid="2AA0833A-AD2B-4866-8153-78C174C13546">
                        <File Id="rgbzip.dll" Source="rgbzip.dll" KeyPath="yes" Checksum="yes"/>
                    </Component>
                </Directory>
            </Directory>

            <Component Id="Registry.CLSID.default" Guid="177CA507-DDC6-4A2A-B6AE-FB712D24AF7B">
                <RegistryKey Root="HKLM"
                             Key="Software\Classes\CLSID\{C1183A9B-F5C0-4DF3-9C29-709AFDFD7FB2}">
                    <RegistryValue Type="string" Value="rgbzip" KeyPath="yes"/>
                </RegistryKey>
            </Component>
            <Component Id="Registry.CLSID.InProcServer.default" Guid="79243FCD-C335-42EA-86C8-9F024B108167">
                <RegistryKey Root="HKLM"
                             Key="Software\Classes\CLSID\{C1183A9B-F5C0-4DF3-9C29-709AFDFD7FB2}\InprocServer32">
                    <RegistryValue Type="string" Value="[INSTALLDIR]rgbzip.dll" KeyPath="yes"/>
                </RegistryKey>
            </Component>
            <Component Id="Registry.CLSID.InProcServer.ThreadingModel" Guid="445B2712-3E17-440E-B317-C81A3A044BE4">
                <RegistryKey Root="HKLM"
                             Key="Software\Classes\CLSID\{C1183A9B-F5C0-4DF3-9C29-709AFDFD7FB2}\InprocServer32">
                    <RegistryValue Type="string" Name="ThreadingModel" Value="Apartment" KeyPath="yes"/>
                </RegistryKey>
            </Component>
            <Component Id="Registry.CLSID.DefaultIcon.default" Guid="76A2F27A-46E3-465D-A242-51D23EDA40FE">
                <RegistryKey Root="HKLM"
                             Key="Software\Classes\CLSID\{C1183A9B-F5C0-4DF3-9C29-709AFDFD7FB2}\DefaultIcon">
                    <RegistryValue Type="expandable" Value="%SystemRoot%\system32\zipfldr.dll" KeyPath="yes"/>
                </RegistryKey>
            </Component>
            <Component Id="Registry.CLSID.ShellFolder.Attributes" Guid="AB755DDF-2642-4E62-908B-A16E834BAD7F">
                <RegistryKey Root="HKLM"
                             Key="Software\Classes\CLSID\{C1183A9B-F5C0-4DF3-9C29-709AFDFD7FB2}\ShellFolder">
                    <RegistryValue Type="integer" Name="Attributes" Value="XA0000020" KeyPath="yes"/>
                </RegistryKey>
            </Component>
        </Directory>

        <Feature Id="DefaultFeature">
            <ComponentRef Id="ShellExtensionDll"/>
            <ComponentRef Id="Registry.CLSID.default"/>
            <ComponentRef Id="Registry.CLSID.InProcServer.default"/>
            <ComponentRef Id="Registry.CLSID.InProcServer.ThreadingModel"/>
            <ComponentRef Id="Registry.CLSID.DefaultIcon.default"/>
            <ComponentRef Id="Registry.CLSID.ShellFolder.Attributes"/>
        </Feature>
    </Product>
</Wix>

```

</details>

**Comment by kennykerr** [comment]
Here's a simple example:

https://github.com/microsoft/windows-rs/blob/master/crates/tests/component/src/lib.rs

**Comment by rbrownwsws** [comment]
Thanks for the help.

Unfortunately I have tried switching out `ManuallyDrop` and `xxx.as_raw()` for `mem::transmute()` as is used in that example and it still does not work.
I get exactly the same behaviour as before. When I open the junction point windows rapidly calls `DllGetClassObject()` about 80 times but then never seems to use or dispose of any of the objects I return.

Can you think of anything else I might be doing wrong? I'm a beginner with COM and am at a bit of a loss as I cannot attach a debugger to windows to see where it is going wrong.

<details>
<summary>DllGetClassObject()</summary>

```rust
/// See [`DllGetClassObject()`](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-dllgetclassobject).
#[no_mangle]
pub extern "system" fn DllGetClassObject(
    rclsid: *const GUID,
    riid: *const GUID,
    ppv: *mut *mut c_void,
) -> HRESULT {
    debug!("DllGetClassObject: Called");

    let rclsid = &unsafe { *rclsid };
    let riid = &unsafe { *riid };
    let ppv = &mut unsafe { *ppv };

    // ppv must be null if we fail so set it here for safety
    *ppv = std::ptr::null_mut();

    if *rclsid != CLSID_RGBZIP {
        warn!(
            "DllGetClassObject: Unexpected CLSID requested: {}",
            guid_to_string(rclsid),
        );

        return CLASS_E_CLASSNOTAVAILABLE;
    }

    if *riid != IID_I_CLASS_FACTORY {
        warn!(
            "DllGetClassObject: Unexpected IID requested: {}",
            guid_to_string(riid),
        );

        return E_UNEXPECTED;
    }

    let factory = RgbzipClassFactory::new(DLL_REF_COUNT.clone());
    let factory: IClassFactory = factory.into();

    *ppv = unsafe { core::mem::transmute(factory) };

    debug!("DllGetClassObject: Success: {:p}", *ppv);

    S_OK
}
```

</details>

**Comment by kennykerr** [comment]
COM can be challenging to get just right. You definitely want to start with a minimal working example that you can debug and build on. Best of luck. For general programming questions I'd suggest https://stackoverflow.com/

**Comment by rbrownwsws** [comment]
The annoying thing is I was trying to go for a minimal example but I'm stumped if I cannot even get windows to use my first COM object.

I guess it would be simpler for me to go for C++ so I at least have a working base example to twiddle with instead of trying to create it from scratch in Rust.

Thanks for your help anyway @kennykerr .

**Comment by kennykerr** [comment]
Yes, it helps to start with client and/or server in C++ (or whatever language you're most comfortable with) so you have a working baseline and then start porting over bit by bit. There's also a test client for the test component I mentioned before:

https://github.com/microsoft/windows-rs/blob/master/crates/tests/component_client/src/lib.rs

**Comment by rbrownwsws** [comment]
I figured it out what the problem was! I thought I would put the solution here in case someone else stumbles into it in the future.

I made a small C++ program that called my DLL so I could try to debug what was happening. I noticed that weirdly `ppv` always remained `NULL` after calling my `DllGetClassObject` method, even though my logging showed it was being set to something else.

It turns out that:

```rust
let ppv = &mut unsafe { *ppv };
```

does not mean the same thing as:

```rust
let ppv = unsafe { &mut *ppv };
```

It turns out that the first version with `&mut` outside the `unsafe` does not just convert the pointer into a reference.
It seems to:
- dereference `ppv`
- copy that value into a hidden variable
- take a reference to that hidden variable
So if you later try to write to `ppv` you are actually writing to that hidden variable, not to where `ppv` originally pointed.

Now that I've figured this out `core::mem::transmute()` seems to be working for me fine.
